phd-tests/tests/src/hw.rs

```diff
@@ -6,22 +6,35 @@ use phd_framework::lifecycle::Action;
 use phd_testcase::*;
 
 #[phd_testcase]
-fn lspci_lifecycle_test(ctx: &Framework) {
+async fn lspci_lifecycle_test(ctx: &Framework) {
     const LSPCI: &str = "sudo lspci -vvx";
     const LSHW: &str = "sudo lshw -notime";
 
-    let mut vm =
-        ctx.spawn_vm(&ctx.vm_config_builder("lspci_lifecycle_test"), None)?;
+    let mut vm = ctx
+        .spawn_vm(&ctx.vm_config_builder("lspci_lifecycle_test"), None)
+        .await?;
 
-    vm.launch()?;
-    vm.wait_to_boot()?;
+    vm.launch().await?;
+    vm.wait_to_boot().await?;
 
-    let lspci = vm.run_shell_command(LSPCI)?;
-    let lshw = vm.run_shell_command(LSHW)?;
-    ctx.lifecycle_test(vm, &[Action::StopAndStart], |vm| {
-        let new_lspci = vm.run_shell_command(LSPCI).unwrap();
-        assert_eq!(new_lspci, lspci);
-        let new_lshw = vm.run_shell_command(LSHW).unwrap();
-        assert_eq!(new_lshw, lshw);
-    })?;
+    let lspci = vm.run_shell_command(LSPCI).await?;
+    let lshw = vm.run_shell_command(LSHW).await?;
+    ctx.lifecycle_test(
+        vm,
+        &[Action::StopAndStart],
+        // REVIEW(gjc): the clone here is very ugly but was the most
+        // straightforward way to convince the compiler that the lspci/lshw
+        // outputs live long enough. There has to be a better way.
+        move |vm| {
+            let lspci = lspci.clone();
+            let lshw = lshw.clone();
+            Box::pin(async move {
```

Comment:
I tried some variations of this before settling on cloning just to get everything compiling so I could get the PR out the door :) 

The basic problem with not moving anything seems to be that rustc is unconvinced that the futures produced by the closure won't outlive the captured references to the caller's locals:

```
28 |           |vm| {
   |           ---- value captured here
29 | /             Box::pin(async {
30 | |                 let new_lspci = vm.run_shell_command(LSPCI).await.unwrap();
31 | |                 assert_eq!(new_lspci, lspci);
   | |                                       ^^^^^ borrowed value does not live long enough
32 | |                 let new_lshw = vm.run_shell_command(LSHW).await.unwrap();
33 | |                 assert_eq!(new_lshw, lshw);
34 | |             })
   | |______________- returning this value requires that `lspci` is borrowed for `'static`
```

One thing that does seem to work is to hand-roll the closure:

- Define `#[async_trait] trait CheckVm` with `async fn check(&self, vm: &TestVm)`
- Change `lifecycle_test` to take an `impl CheckVm`
- In the test cases, define a struct that contains all of the locals needed to perform the check, then `impl CheckVm` for it and pass an instance of it to `lifecycle_test`

I'm a little fuzzy on how this works, but AIUI what ends up happening here is that the check function ends up being expanded to

`fn check<'context, 'v>(&'context self, vm: &'v TestVm) -> Pin<Box<dyn Future<Output = ()> + 'context>>`

which (again AIUI) works because now we've assigned a lifetime to the "captured variables" (i.e. the stuff we put into `self`) that we can refer to when we name the lifetime of the future returned by the trait function.

This is part of the reason for the comment I left here--I can figure out how to express that the future can't outlive the lifetime of the *test VM* declared in the higher-rank trait bound, i.e. I could write out the `BoxFuture` as

`check_fn: F ... where F: for<'v> Fn(&'v TestVm) -> Pin<Box<dyn Future<Output = ()> + 'v`

...but it's not clear to me how to name the lifetime of the closed-over variables in the caller.

---

I think the tradeoff here between extra verbosity (defining the structs and impls to hand-roll the closures) and the awkwardness of having to move and clone everything is a very close one. I committed the hand-rolled variation in ff41adce to give everyone a look at what it looks like. I might also ask in our Rust chat channel if there's a way around all this that I've missed.

Reply:
Hmm, the async-trait approach is...interesting. IMO, the need to have to define a struct and a trait impl for it is somewhat more annoying than cloning into the closure, since we still need to hand-implement the async-trait struct even when the closure _doesn't_ capture anything. But, I'm fine with this either way, honestly.

Reply:
I'm still unsatisfied enough with this that I'm going to noodle on it for a bit longer before merging anything :)

Reply:
I played around with some other ideas here, some of them of dubious quality ("what if we used a macro to construct the lifecycle testing loop?!") before looking up and realizing there were higher-leverage ways for me to spend my time :)

I agree with your comment here--having to clone is annoying but the trait object thing is more annoying. I went back to the original approach (and cleaned up the `REVIEW` comments) in 1bb7bcb.